File: sdk/java/nmidsdk/worker/src/main/java/com/nmidsdk/worker/Agent.java
package com.nmidsdk.worker;

import com.nmidsdk.model.Constants;
import com.nmidsdk.model.utils.Utils;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Deque;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Worker agent: a single connection from a worker to one nmid server.
 * Mirrors sdk/node/lib/worker/agent.js, but uses a blocking socket with a
 * dedicated reader thread.
 */
public class Agent {
    public final String network;
    private final Object addr;
    public final Worker worker;

    public Socket conn;
    private OutputStream out;
    private InputStream in;

    public final Request req = new Request();
    public volatile long lastTime;

    private volatile boolean closed = false;
    private Thread readerThread;

    // FIFO of resolvers waiting for a PDT_OK confirmation from the server.
    private final Deque<CompletableFuture<Void>> okResolvers = new ConcurrentLinkedDeque<>();

    public Agent(String network, Object addr, Worker worker) {
        this.network = network == null ? "tcp" : network;
        this.addr = addr;
        this.worker = worker;
        this.lastTime = Utils.getMillisecond();
    }

    /**
     * Wait for the next PDT_OK from the server. The server sends PDT_OK in
     * response to PDT_W_SET_NAME and PDT_W_ADD_FUNC, so awaiting this after
     * each registration packet guarantees the server has actually processed it
     * before we proceed (closes the "have no job do" race for early client
     * calls). A safety timeout prevents hanging forever if the OK is lost.
     */
    public CompletableFuture<Void> expectOk(long timeoutMs) {
        CompletableFuture<Void> future = new CompletableFuture<>();
        okResolvers.add(future);

        ScheduledExecutorService scheduler = worker.getScheduler();
        if (scheduler != null) {
            scheduler.schedule(() -> {
                if (!future.isDone()) {
                    okResolvers.remove(future);
                    worker.emitWarning(new Exception("PDT_OK timeout, proceeding anyway"));
                    future.complete(null);
                }
            }, timeoutMs, TimeUnit.MILLISECONDS);
        }
        return future;
    }

    public void onOk() {
        CompletableFuture<Void> f = okResolvers.poll();
        if (f != null) {
            f.complete(null);
        }
    }

    private String[] parseAddr() {
        if (addr instanceof String[]) {
            String[] a = (String[]) addr;
            if (a.length < 2) throw new IllegalArgumentException("invalid addr array");
            return new String[]{a[0], a[1]};
        }
        if (addr instanceof String) {
            String s = (String) addr;
            int idx = s.lastIndexOf(':');
            if (idx == -1) throw new IllegalArgumentException("invalid addr: " + s);
            return new String[]{s.substring(0, idx), s.substring(idx + 1)};
        }
        throw new IllegalArgumentException("invalid addr: " + addr);
    }

    /** Connect to the server. Completes on success. */
    public CompletableFuture<Void> connect() {
        CompletableFuture<Void> future = new CompletableFuture<>();
        try {
            String[] hp = parseAddr();
            conn = new Socket();
            conn.connect(new InetSocketAddress(hp[0], Integer.parseInt(hp[1])), Constants.DIAL_TIME_OUT);
            conn.setTcpNoDelay(true);
            out = conn.getOutputStream();
            in = conn.getInputStream();
            closed = false;
            lastTime = Utils.getMillisecond();

            readerThread = new Thread(this::readerLoop, "nmid-agent-reader");
            readerThread.setDaemon(true);
            readerThread.start();

            future.complete(null);
        } catch (Exception e) {
            future.completeExceptionally(e);
        }
        return future;
    }

    /** Reconnect an existing agent after a connection drop. */
    public CompletableFuture<Void> reConnect() {
        if (conn != null) {
            try { conn.close(); } catch (Exception ignored) {}
            conn = null;
        }
        return connect();
    }

    private void readerLoop() {
        try {
            while (!closed) {
                byte[] header = readN(in, Constants.MIN_DATA_SIZE);
                if (header == null) break;

                int connType = Response.readIntBE(header, 0);
                int dataType = Response.readIntBE(header, 4);
                int dataLen = Response.readIntBE(header, 8);

                byte[] body = readN(in, dataLen);
                if (body == null) break;

                if (connType != Constants.CONN_TYPE_SERVER) continue;

                Response resp = Response.decode(dataType, dataLen, body);
                if (resp == null) continue;
                resp.agent = this;

                // Dispatch on the worker (reader thread is single per agent, so
                // ordering is preserved; job execution itself is offloaded).
                worker.handleResp(resp);
            }
        } catch (IOException e) {
            if (!closed) {
                worker.emitError(e);
            }
        }
    }

    // ---- atomic send helpers ------------------------------------------------
    // Each method builds the packet and writes it under the agent lock, so
    // concurrent writers (reader thread returning a job, heartbeat timer,
    // registration) never corrupt the shared Request.

    public synchronized void sendSetName(String name) {
        req.setWorkerName(name);
        rawWrite();
    }

    public synchronized void sendAddFunc(String name) {
        req.addFunctionPack(name);
        rawWrite();
    }

    public synchronized void sendDelFunc(String name) {
        req.delFunctionPack(name);
        rawWrite();
    }

    public synchronized void sendHeartbeat() {
        req.heartBeatPack();
        rawWrite();
    }

    public synchronized void sendGrabJob() {
        req.grabDataPack();
        rawWrite();
    }

    public synchronized void sendWakeup() {
        req.wakeupPack();
        rawWrite();
    }

    public synchronized void sendReturnData(Response job, byte[] ret) {
        req.handleLen = job.handleLen;
        req.handle = job.handle;
        req.paramsLen = job.paramsLen;
        req.params = job.params;
        req.jobIdLen = job.jobIdLen;
        req.jobId = job.jobId;
        req.retPack(ret);
        rawWrite();
    }

    private void rawWrite() {
        if (conn == null || closed) return;
        byte[] buf = req.encodePack();
        try {
            out.write(buf);
            out.flush();
        } catch (IOException e) {
            worker.emitError(e);
        }
    }

    public void heartBeatPing() {
        sendHeartbeat();
    }

    public void delOldFuncMsg(String funcName) {
        sendDelFunc(funcName);
    }

    public void wakeup() {
        sendWakeup();
    }

    public void close() {
        closed = true;
        Socket s = conn;
        if (s != null) {
            try { s.close(); } catch (Exception ignored) {}
            conn = null;
        }
    }

    /** Read exactly n bytes, blocking; returns null on EOF. */
    private static byte[] readN(InputStream in, int n) throws IOException {
        if (n == 0) return new byte[0];
        byte[] buf = new byte[n];
        int off = 0;
        while (off < n) {
            int r = in.read(buf, off, n - off);
            if (r < 0) return null;
            off += r;
        }
        return buf;
    }
}

File: sdk/node/lib/worker/agent.js
'use strict';

// Worker agent: a single connection from a worker to one nmid server.
// Mirrors sdk/python3/nmidsdk/worker/agent.py, but uses Node's event-driven
// socket so no read thread is required.

const net = require('net');
const { EventEmitter } = require('events');
const Request = require('./request');
const { decodePack } = require('./response');
const { getMillisecond } = require('./utils');
const C = require('../const');

class Agent extends EventEmitter {
  // worker: the owning Worker instance
  constructor(network, addr, worker) {
    super();
    this.network = network || 'tcp';
    this.addr = addr; // [host, port] or 'host:port'
    this.worker = worker;
    this.conn = null;
    this.req = new Request();
    this.lastTime = getMillisecond();
    this.recvBuffer = Buffer.alloc(0);
    this._closed = false;

    // FIFO of resolvers waiting for a PDT_OK confirmation from the server.
    // Populated by expectOk(), drained by onOk().
    this._okResolvers = [];
  }

  // Wait for the next PDT_OK from the server. The server sends PDT_OK in
  // response to PDT_W_SET_NAME and PDT_W_ADD_FUNC, so awaiting this after each
  // registration packet guarantees the server has actually processed it before
  // we proceed (closes the "have no job do" race for early client calls).
  // A safety timeout prevents hanging forever if the OK is ever lost.
  expectOk(timeoutMs = 5000) {
    return new Promise((resolve) => {
      let done = false;
      const finish = () => {
        if (done) return;
        done = true;
        resolve();
      };
      this._okResolvers.push(finish);
      setTimeout(() => {
        if (!done) {
          const idx = this._okResolvers.indexOf(finish);
          if (idx !== -1) this._okResolvers.splice(idx, 1);
          if (this.worker) this.worker.emit('warning', new Error('PDT_OK timeout, proceeding anyway'));
          finish();
        }
      }, timeoutMs);
    });
  }

  onOk() {
    const finish = this._okResolvers.shift();
    if (finish) finish();
  }

  _parseAddr(addr) {
    if (Array.isArray(addr)) {
      return { host: addr[0], port: addr[1] };
    }
    if (typeof addr === 'string') {
      const idx = addr.lastIndexOf(':');
      if (idx === -1) throw new Error('invalid addr: ' + addr);
      return { host: addr.slice(0, idx), port: parseInt(addr.slice(idx + 1), 10) };
    }
    throw new Error('invalid addr: ' + addr);
  }

  // Connect to the server. Resolves on success, rejects on failure.
  connect() {
    return new Promise((resolve, reject) => {
      const { host, port } = this._parseAddr(this.addr);
      const socket = net.createConnection({ host, port });
      this.conn = socket;
      this._closed = false;

      let settled = false;
      socket.setTimeout(C.DIAL_TIME_OUT * 1000);
      socket.once('connect', () => {
        socket.setTimeout(0);
        settled = true;
        this.lastTime = getMillisecond();
        this.emit('connect');
        resolve();
      });
      socket.once('timeout', () => {
        if (!settled) {
          settled = true;
          socket.destroy(new Error('connect timeout'));
          reject(new Error('connect timeout'));
        }
      });
      socket.once('error', (err) => {
        if (!settled) {
          settled = true;
          reject(err);
        }
      });

      socket.on('data', (chunk) => this._onData(chunk));
      socket.on('error', (err) => this._onError(err));
      socket.on('close', () => this._onClose());
    });
  }

  // Reconnect an existing agent after a connection drop.
  reConnect() {
    return new Promise((resolve, reject) => {
      if (this.conn) {
        try { this.conn.destroy(); } catch (e) { /* ignore */ }
        this.conn = null;
      }
      const { host, port } = this._parseAddr(this.addr);
      const socket = net.createConnection({ host, port });
      this.conn = socket;
      this._closed = false;

      socket.setTimeout(C.DIAL_TIME_OUT * 1000);
      socket.once('connect', () => {
        socket.setTimeout(0);
        this.lastTime = getMillisecond();
        resolve();
      });
      socket.once('timeout', () => {
        socket.destroy(new Error('connect timeout'));
        reject(new Error('connect timeout'));
      });
      socket.once('error', reject);

      socket.on('data', (chunk) => this._onData(chunk));
      socket.on('error', (err) => this._onError(err));
      socket.on('close', () => this._onClose());
    });
  }

  _onData(chunk) {
    this.recvBuffer = this.recvBuffer.length
      ? Buffer.concat([this.recvBuffer, chunk])
      : chunk;
    this._processBuffer();
  }

  _processBuffer() {
    while (true) {
      if (this.recvBuffer.length < C.MIN_DATA_SIZE) return;
      const dataLen = this.recvBuffer.readUInt32BE(8);
      const packetLen = C.MIN_DATA_SIZE + dataLen;
      if (this.recvBuffer.length < packetLen) return;

      const packet = this.recvBuffer.subarray(0, packetLen);
      this.recvBuffer = this.recvBuffer.subarray(packetLen);

      const { resp, err } = decodePack(packet);
      if (err || !resp) continue;

      resp.agent = this;
      // Dispatch directly on the worker (single event loop, no queue needed).
      this.worker.handleResp(resp);
    }
  }

  write() {
    if (!this.conn || this._closed) return;
    const buf = this.req.encodePack();
    this.conn.write(buf);
  }

  heartBeatPing() {
    this.req.heartBeatPack();
    this.write();
  }

  grab() {
    this.req.grabDataPack();
    this.write();
  }

  wakeup() {
    this.req.wakeupPack();
    this.write();
  }

  limitExceed() {
    this.req.limitExceedPack();
    this.write();
  }

  delOldFuncMsg(funcName) {
    this.req.delFunctionPack(funcName);
    this.write();
  }

  reAddFuncMsg(funcName) {
    this.req.addFunctionPack(funcName);
    this.write();
  }

  reSetWorkerName(workerName) {
    this.req.setWorkerName(workerName);
    this.write();
  }

  _onError(err) {
    this.emit('error', err);
  }

  _onClose() {
    this.conn = null;
    this.emit('close');
  }

  close() {
    this._closed = true;
    if (this.conn) {
      try { this.conn.destroy(); } catch (e) { /* ignore */ }
      this.conn = null;
    }
  }
}

module.exports = Agent;
